ons-spark/ancillary-topics/pandas_udfs_rescue.ipynb

```diff
@@ -0,0 +1,902 @@
+       "      <td>Dog</td>\n",
+       "      <td>1.0</td>\n",
+       "      <td>255.0</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>8</th>\n",
+       "      <td>4306091</td>\n",
+       "      <td>Squirrel</td>\n",
+       "      <td>1.0</td>\n",
+       "      <td>255.0</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>9</th>\n",
+       "      <td>4715091</td>\n",
+       "      <td>Dog</td>\n",
+       "      <td>1.0</td>\n",
+       "      <td>255.0</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "  incident_number                      animal_group  incident_duration  \\\n",
+       "0          139091                               Dog                2.0   \n",
+       "1          275091                               Fox                1.0   \n",
+       "2         2075091                               Dog                1.0   \n",
+       "3         2872091                             Horse                1.0   \n",
+       "4         3553091                            Rabbit                1.0   \n",
+       "5         3742091  Unknown - Heavy Livestock Animal                1.0   \n",
+       "6         4011091                               Dog                1.0   \n",
+       "7         4211091                               Dog                1.0   \n",
+       "8         4306091                          Squirrel                1.0   \n",
+       "9         4715091                               Dog                1.0   \n",
+       "\n",
+       "   incident_cost  \n",
+       "0          510.0  \n",
+       "1          255.0  \n",
+       "2          255.0  \n",
+       "3          255.0  \n",
+       "4          255.0  \n",
+       "5          255.0  \n",
+       "6          255.0  \n",
+       "7          255.0  \n",
+       "8          255.0  \n",
+       "9          255.0  "
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Check data looks as it should\n",
+    "\n",
+    "rescue.limit(10).toPandas()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Scalar Pandas UDFs\n",
+    "\n",
+    "Scalar Pandas UDFs take in a `pandas.Series` and output a `pandas.Series`. \n",
+    "\n",
+    "Spark executes scalar Pandas UDFs by serialising each partition column into a `pandas.Series` object (basically splitting the data into batches). The UDF is then called on each of of these Series objects as a subset of the data and then results are concactenated together and returned as a `pandas.Series`. \n",
+    "\n",
+    "Be aware that scalar UDFs may cause skewed results when calculating means and standard deviations. For more info on this and how to overcome it see [here](https://towardsdatascience.com/pyspark-or-pandas-why-not-both-95523946ec7c).\n",
```

Comment:
I would say the mean and standard deviation is either correct or not, so if it's not correct let's say it's incorrect.

Reply:
> I would say the mean and standard deviation is either correct or not, so if it's not correct let's say it's incorrect.

So this only happens if your df is over multiple partitions, hence the word "may". I have put in brackets if the df is over multiple partitions.